File: lab-3-dynamodb/dynamoDBCSharpLab/Lab3/InfectionsTableCreator.cs
// Copyright 2017 Amazon Web Services, Inc. or its affiliates. All rights reserved.

using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.Model;
using Amazon.Runtime;
using System.Collections.Generic;
using System.Diagnostics;

namespace Lab3
{
    // The InfectionsTableCreator class creates a table and global secondary
    // index to store data about infections.
    public static class InfectionsTableCreator
    {
        public static readonly string InfectionsTableName = "Infections";
        public static readonly string CityDateIndexName = "InfectionsByCityDate";
        public static AmazonDynamoDBClient dynamoDBClient = null;

        // Entry
        public static void main()
        {
            Init();

            // In this sample, we will use the Document API to create our
            // designed DynamoDB table.
            try
            {
                RemoveInfectionsTableIfExists();
                CreateInfectionsTable();
            }
            catch (AmazonServiceException ase)
            {
                /*
                * AmazonServiceException represents an error response from an
                * AWS service.  AWS service received the request but either
                * found it invalid or encountered an error trying to execute it.
                */
                Debug.WriteLine("Error Message:" + ase.Message);
                Debug.WriteLine("HTTP Status Code:" + ase.StatusCode);
                Debug.WriteLine("AWS Error Code:" + ase.ErrorCode);
                Debug.WriteLine("Error Type:" + ase.ErrorType);
                Debug.WriteLine("Request ID:" + ase.RequestId);
            }
            catch (AmazonClientException ace)
            {
                /*
                 * AmazonClientException represents an error that occurred
                 * inside the client on the local host, either while trying to
                 * send the request to AWS or interpret the response.
                 * For example, if no network connection is available, the
                 * client won't be able to connect to AWS to execute a request
                 * and will throw an AmazonClientException.
                 */
                Debug.WriteLine("Error Message:" + ace.Message);
            }
        }

        private static void CreateInfectionsTable()
        {
            // Create an instance of a GlobalSecondaryIndex class
            GlobalSecondaryIndex gsi = new GlobalSecondaryIndex
            {
                IndexName = CityDateIndexName,
                ProvisionedThroughput = new ProvisionedThroughput
                {
                    ReadCapacityUnits = 5L,
                    WriteCapacityUnits = 5L
                },
                Projection = new Projection { ProjectionType = "ALL" }
            };

            var indexKeySchema = new List<KeySchemaElement>
            {
                { new KeySchemaElement { AttributeName = "City", KeyType = "HASH"} },
                { new KeySchemaElement{AttributeName = "Date",KeyType = "RANGE"} },
            };
            gsi.KeySchema = indexKeySchema;

            // Create attribute definitions for PatientId, City, Date.
            var attributeDefinitions = new List<AttributeDefinition>()
            {
                {
                    new AttributeDefinition {
                        AttributeName = "PatientID",
                        AttributeType = "S",
                    }
                },
                {
                    new AttributeDefinition {
                        AttributeName = "City",
                        AttributeType = "S",
                    }
                },
                {
                    new AttributeDefinition() {
                        AttributeName = "Date",
                        AttributeType = "S",
                    }
                },
            };
            // Table key schema
            var tableKeySchema = new List<KeySchemaElement>()
            {
                {
                    new KeySchemaElement {
                        AttributeName = "PatientID",
                        KeyType = "HASH"
                    }
                }
            };

            Debug.WriteLine("Creating DynamoDB table.");
            CreateInfectionsTableWithIndex(attributeDefinitions, tableKeySchema, gsi);

            string status = null;
            do
            {
                status = GetTableStatus();
            } while (status != TableStatus.ACTIVE);

            Debug.WriteLine("Table successfully created.");
        }

        private static void RemoveInfectionsTableIfExists()
        {
            try
            {
                string tableName = InfectionsTableName;
                Debug.WriteLine("Attempting to delete DynamoDB table if one already exists.");

                var request = new DeleteTableRequest { TableName = tableName };
                dynamoDBClient.DeleteTable(request);
            }
            catch (ResourceNotFoundException e)
            {
                Debug.Write("{0} table does not exist. Do not need to remove it. \n", InfectionsTableName + " " + e.Message);
            }
        }

        private static void Init()
        {
            dynamoDBClient = new AmazonDynamoDBClient();
        }

        /**
         * Check the table's status using the DescribeTable method
         * Wait for the table to become active
         *
         * @return      Table Status
         */
        private static string GetTableStatus()
        {
            // STUDENT TODO 1: Replace the solution with your own code
            return Solution.GetTableStatus(dynamoDBClient, InfectionsTableName);
        }

        /**
         * Create an instance of CreateTableRequest class
         * Create table using the request defined
         *
         * @param attributeDefinitions    Attribute definitions for the table
         * @param tableKeySchema          Table's key schema
         * @param gsi                     Global secondary index
         */
        private static void CreateInfectionsTableWithIndex(List<AttributeDefinition> attributeDefinitions, List<KeySchemaElement> tableKeySchema, GlobalSecondaryIndex gsi)
        {
            // STUDENT TODO 2: Replace the solution with your own code
            Solution.CreateInfectionsTableWithIndex(dynamoDBClient, InfectionsTableName, attributeDefinitions, tableKeySchema, gsi);
        }
    }
}

File: lab-3-dynamodb/dynamoDBJavaLab/src/InfectionsTableCreator.java
// Copyright 2017 Amazon Web Services, Inc. or its affiliates. All rights reserved.

import com.amazonaws.AmazonClientException;
import com.amazonaws.AmazonServiceException;
import com.amazonaws.regions.Region;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient;
import com.amazonaws.services.dynamodbv2.document.DynamoDB;
import com.amazonaws.services.dynamodbv2.document.Table;
import com.amazonaws.services.dynamodbv2.model.AttributeDefinition;
import com.amazonaws.services.dynamodbv2.model.DescribeTableResult;
import com.amazonaws.services.dynamodbv2.model.GlobalSecondaryIndex;
import com.amazonaws.services.dynamodbv2.model.KeySchemaElement;
import com.amazonaws.services.dynamodbv2.model.KeyType;
import com.amazonaws.services.dynamodbv2.model.Projection;
import com.amazonaws.services.dynamodbv2.model.ProjectionType;
import com.amazonaws.services.dynamodbv2.model.ProvisionedThroughput;
import com.amazonaws.services.dynamodbv2.model.ResourceNotFoundException;
import java.util.ArrayList;

// The InfectionsTableCreator class creates a table and global secondary index to store data about infections
// This sample uses the DynamoDB document API
public class InfectionsTableCreator {

  public static final String INFECTIONS_TABLE_NAME = "Infections";
  public static final String CITY_DATE_INDEX_NAME = "InfectionsByCityDate";
  public static final Region REGION = Utils.getRegion();
  public static DynamoDB dynamoDB = null;
  public static AmazonDynamoDBClient dynamoDBClient = null;

  public static void main(String[] args) throws Exception {
    try {
      // Create an instance of AmazonDynamoDBClient class
      dynamoDBClient = new AmazonDynamoDBClient();

      // Set the client's region
      dynamoDBClient.setRegion(REGION);

      // Create an instance of DynamoDB class
      dynamoDB = new DynamoDB(dynamoDBClient);

      // Remove the table if it already exists
      removeInfectionsTableIfExists();

      // Create the infections table and index
      createInfectionsTableWrapper();

    } catch (AmazonServiceException ase) {
      System.out.println("Error Message:    " + ase.getMessage());
      System.out.println("HTTP Status Code: " + ase.getStatusCode());
      System.out.println("AWS Error Code:   " + ase.getErrorCode());
      System.out.println("Error Type:       " + ase.getErrorType());
      System.out.println("Request ID:       " + ase.getRequestId());
    } catch (AmazonClientException ace) {
      System.out.println("Error Message: " + ace.getMessage());
    }
  }

  private static void removeInfectionsTableIfExists() {
    try {
      Table table = dynamoDB.getTable(INFECTIONS_TABLE_NAME);
      DescribeTableResult descTableResult = dynamoDBClient.describeTable(INFECTIONS_TABLE_NAME);
      if (descTableResult != null && descTableResult.getTable().getTableStatus().equals("ACTIVE")) {
        System.out.println("Deleting table");
        table.delete();
        table.waitForDelete();
      }
    } catch (ResourceNotFoundException e) {
      System.out.printf("%s table does not exist \n", INFECTIONS_TABLE_NAME);
    } catch (InterruptedException ie) {
    }
  }

  private static void createInfectionsTableWrapper() {
    // Create attribute definitions for the primary key attributes of the table and indexes
    ArrayList<AttributeDefinition> attributeDefinitions = new ArrayList<AttributeDefinition>();
    attributeDefinitions.add(
        new AttributeDefinition().withAttributeName("PatientId").withAttributeType("S"));
    attributeDefinitions.add(
        new AttributeDefinition().withAttributeName("City").withAttributeType("S"));
    attributeDefinitions.add(
        new AttributeDefinition().withAttributeName("Date").withAttributeType("S"));

    // Create key schema element for the table's primary key attribute
    KeySchemaElement tableKeySchemaElem =
        new KeySchemaElement().withAttributeName("PatientId").withKeyType(KeyType.HASH);

    // Create object to specify table's provisioned throughput
    ProvisionedThroughput tableProvisionedThroughput = new ProvisionedThroughput(5L, 10L);

    // Create global secondary index object
    // The code uses fluent setter methods to initialize the GlobalSecondaryIndex object
    GlobalSecondaryIndex gsi =
        new GlobalSecondaryIndex()
            .withIndexName(CITY_DATE_INDEX_NAME)
            .withKeySchema(
                new KeySchemaElement("City", KeyType.HASH),
                new KeySchemaElement("Date", KeyType.RANGE))
            .withProvisionedThroughput(new ProvisionedThroughput(5L, 5L))
            .withProjection(new Projection().withProjectionType(ProjectionType.ALL));

    // Create the table and index by using the given parameters
    createInfectionsTableWithIndex(
        dynamoDB,
        INFECTIONS_TABLE_NAME,
        attributeDefinitions,
        tableKeySchemaElem,
        tableProvisionedThroughput,
        gsi);
  }

  /**
   * Create a DynamoDB table
   *
   * @param dynamoDB                    Instance of DynamoDB class
   * @param tableName                   Table name
   * @param attributeDefinitions        Attribute definitions for the table
   * @param tableKeySchemaElem          Table's key schema element
   * @param tableProvisionedThroughput  Provisioned Throughput
   * @param gsi                         Global secondary index
   */
  public static void createInfectionsTableWithIndex(
      DynamoDB dynamoDB,
      String tableName,
      ArrayList<AttributeDefinition> attributeDefinitions,
      KeySchemaElement tableKeySchemaElem,
      ProvisionedThroughput tableProvisionedThroughput,
      GlobalSecondaryIndex gsi) {
    // STUDENT TODO 1: Replace the solution with your own code
    Solution.createInfectionsTableWithIndex(
        dynamoDB,
        INFECTIONS_TABLE_NAME,
        attributeDefinitions,
        tableKeySchemaElem,
        tableProvisionedThroughput,
        gsi);
  }
}
